File: source/packages/my/HistoryComboBox.pas
unit HistoryComboBox;

interface

uses
  Windows, Messages, SysUtils, Classes, Graphics, Controls, Forms, Dialogs,
  StdCtrls;

type
  TUIStateAwareComboBox = class(TComboBox)
  private
    FDarkMode: Boolean;
    procedure SetDarkMode(Value: Boolean);
  protected
    procedure ComboWndProc(var Message: TMessage; ComboWnd: HWnd; ComboProc: TWindowProcPtr); override;
    procedure CreateWnd; override;
  public
    property DarkMode: Boolean read FDarkMode write SetDarkMode default False;
  end;

type
  THistorySaveOn = set of (soExit, soDropDown);

const
  DefaultHistorySaveOn = [soExit, soDropDown];
  DefaultMaxHistorySize = 30;

type
  THistoryComboBox = class;

  THistoryComboBoxGetData = procedure(Sender: THistoryComboBox; var Data: Pointer) of object;
  THistoryComboBoxSetData = procedure(Sender: THistoryComboBox; Data: Pointer) of object;

  THistoryComboBox = class(TUIStateAwareComboBox)
  private
    { Private declarations }
    FSaveOn: THistorySaveOn;
    FMaxHistorySize: Integer;
    FOnGetData: THistoryComboBoxGetData;
    FOnSetData: THistoryComboBoxSetData;

    procedure SetMaxHistorySize(AMaxHistorySize: Integer);
    function GetMaxItemWidth: Integer;
  protected
    { Protected declarations }
    procedure DoExit; override;
    procedure DropDown; override;
    procedure KeyDown(var Key: Word; Shift: TShiftState); override;
    procedure Change; override;
  public
    { Public declarations }
    constructor Create(AOwner: TComponent); override;
    procedure SaveToHistory; virtual;
  published
    { Published declarations }
    property SaveOn: THistorySaveOn read FSaveOn write FSaveOn default DefaultHistorySaveOn;
    property MaxHistorySize: Integer read FMaxHistorySize write SetMaxHistorySize default DefaultMaxHistorySize;
    property OnGetData: THistoryComboBoxGetData read FOnGetData write FOnGetData;
    property OnSetData: THistoryComboBoxSetData read FOnSetData write FOnSetData;
  end;

procedure SaveToHistory(Strings: TStrings; T: string; Data: Pointer = nil; MaxHistorySize: Integer = DefaultMaxHistorySize);

procedure Register;

implementation

uses
  PasTools;

procedure Register;
begin
  RegisterComponents('Martin', [THistoryComboBox]);
end;

procedure SaveToHistory(Strings: TStrings; T: string; Data: Pointer; MaxHistorySize: Integer);
begin
  if T <> '' then
  begin
    while Strings.IndexOf(T) >= 0 do Strings.Delete(Strings.IndexOf(T));
    Strings.InsertObject(0, T, TObject(Data));
  end;
  while Strings.Count > MaxHistorySize do
    Strings.Delete(Strings.Count-1);
end;

  { TUIStateAwareComboBox }

procedure TUIStateAwareComboBox.ComboWndProc(var Message: TMessage; ComboWnd: HWnd; ComboProc: TWindowProcPtr);
begin
  inherited;

  if Message.Msg = WM_SYSKEYDOWN then
  begin
    UpdateUIState(TWMKey(Message).CharCode);
  end;
end;

procedure TUIStateAwareComboBox.CreateWnd;
begin
  inherited;
  if DarkMode then
    SetDarkModeTheme(Self, 'CFD');
end;

procedure TUIStateAwareComboBox.SetDarkMode(Value: Boolean);
begin
  if DarkMode <> Value then
  begin
    FDarkMode := Value;
    RecreateWnd;
  end;
end;

  { THistoryComboBox }

constructor THistoryComboBox.Create(AOwner: TComponent);
begin
  inherited;
  FSaveOn := DefaultHistorySaveOn;
  FMaxHistorySize := DefaultMaxHistorySize;
  FOnGetData := nil;
  FOnSetData := nil;
end;

procedure THistoryComboBox.KeyDown(var Key: Word; Shift: TShiftState);
begin
  if ((Key = VK_DOWN) or (Key = VK_UP)) and
     (not (ssAlt in Shift)) and (soDropDown in SaveOn) then
  begin
    if Items.IndexOf(Text) < 0 then SaveToHistory;
  end;
  if DroppedDown and (Key = VK_DELETE) and (ssCtrl in Shift) and (SaveOn <> []) then
  begin
    Items.Clear;
    Key := 0;
  end;
  inherited;
end;

procedure THistoryComboBox.SetMaxHistorySize(AMaxHistorySize: Integer);
begin
  FMaxHistorySize := AMaxHistorySize;
  while Items.Count > FMaxHistorySize do
    Items.Delete(Items.Count-1);
end;

procedure THistoryComboBox.DoExit;
begin
  inherited;
  if soExit in SaveOn then SaveToHistory;
end;

procedure THistoryComboBox.DropDown;
var
  ItemWidth: Integer;
begin
  inherited;
  if soDropDown in SaveOn then SaveToHistory;

  ItemWidth := GetMaxItemWidth + ScaleByPixelsPerInch(8, Self);
  if Items.Count > DropDownCount then
    Inc(ItemWidth, GetSystemMetricsForControl(Self, SM_CXVSCROLL));
  Self.Perform(CB_SETDROPPEDWIDTH, ItemWidth, 0);
end;

procedure THistoryComboBox.Change;
var
  Index: Integer;
begin
  inherited Change;
  if Assigned(OnSetData) then
  begin
    // note that ItemIndex is not reliable
    Index := Items.IndexOf(Text);
    if Index >= 0 then OnSetData(Self, Items.Objects[Index]);
  end;
end;

procedure THistoryComboBox.SaveToHistory;
var
  Data: Pointer;
begin
  if Text <> '' then
  begin
    Data := nil;
    if Assigned(OnGetData) then
      OnGetData(Self, Data);
    HistoryComboBox.SaveToHistory(Items, Text, Data, MaxHistorySize);
    ItemIndex := 0;
  end;
end;

function THistoryComboBox.GetMaxItemWidth: Integer;
var
  DC: HDC;
  SaveFont: HFont;
  Size: TSize;
  Index: Integer;
begin
  Result := 0;
  DC := GetDC(0);
  try
    SaveFont := SelectObject(DC, Font.Handle);
    for Index := 0 to Items.Count - 1 do
    begin
      GetTextExtentPoint32(DC, PChar(Items[Index]), Length(Items[Index]), Size);
      if Size.Cx > Result then Result := Size.Cx;
    end;
    SelectObject(DC, SaveFont);
  finally
    ReleaseDC(0, DC);
  end;
end;

initialization
end.

File: source/packages/my/HistoryComboBox.cpp
//---------------------------------------------------------------------------
#include <vcl.h>
#pragma hdrstop

#include "HistoryComboBox.h"
#include "PasTools.hpp"
#include "Common.h"
#include <algorithm>
//---------------------------------------------------------------------------
#pragma package(smart_init)
//---------------------------------------------------------------------------
void SaveToHistory(TStrings * Strings, const UnicodeString & T, TObject * Data, int MaxHistorySize)
{
  if (!T.IsEmpty())
  {
    while (Strings->IndexOf(T) >= 0)
    {
      Strings->Delete(Strings->IndexOf(T));
    }
    Strings->InsertObject(0, T, Data);
  }

  while (Strings->Count > MaxHistorySize)
  {
    Strings->Delete(Strings->Count - 1);
  }
}
//---------------------------------------------------------------------------
// TUIStateAwareComboBox
//---------------------------------------------------------------------------
__fastcall TUIStateAwareComboBox::TUIStateAwareComboBox(TComponent * AOwner) :
  TComboBox(AOwner)
{
}
//---------------------------------------------------------------------------
void __fastcall TUIStateAwareComboBox::ComboWndProc(TMessage & Message, HWND ComboWnd, TWindowProcPtr ComboProc)
{
  TComboBox::ComboWndProc(Message, ComboWnd, ComboProc);

  if (Message.Msg == WM_SYSKEYDOWN)
  {
    UpdateUIState(reinterpret_cast<TWMKey &>(Message).CharCode);
  }
}
//---------------------------------------------------------------------------
void __fastcall TUIStateAwareComboBox::CreateWnd()
{
  TComboBox::CreateWnd();

  if (DarkMode)
  {
    SetDarkModeTheme(this, L"CFD");
  }
}
//---------------------------------------------------------------------------
void TUIStateAwareComboBox::SetDarkMode(bool Value)
{
  if (DarkMode != Value)
  {
    FDarkMode = Value;
    RecreateWnd();
  }
}
//---------------------------------------------------------------------------
// THistoryComboBox
//---------------------------------------------------------------------------
THistoryComboHistoryEvent THistoryComboBox::OnLoadHistory = nullptr;
THistoryComboHistoryEvent THistoryComboBox::OnSaveHistory = nullptr;
//---------------------------------------------------------------------------
__fastcall THistoryComboBox::THistoryComboBox(TComponent * AOwner) :
  TUIStateAwareComboBox(AOwner)
{
  FSaveOn = THistorySaveOn(DefaultHistorySaveOn);
  FMaxHistorySize = DefaultMaxHistorySize;
  FOnGetData = nullptr;
  FOnSetData = nullptr;
  DropDownCount = DefaultHistoryDropDownCount;
  AutoComplete = DefaultHistoryAutoComplete;
}
//---------------------------------------------------------------------------
void __fastcall THistoryComboBox::KeyDown(unsigned short & Key, TShiftState Shift)
{
  if (((Key == VK_DOWN) || (Key == VK_UP)) &&
       !Shift.Contains(ssAlt) && SaveOn.Contains(soDropDown))
  {
    if (Items->IndexOf(Text) < 0)
    {
      AddToHistory();
    }
  }
  if (DroppedDown && (Key == VK_DELETE) && Shift.Contains(ssCtrl) && !SaveOn.Empty())
  {
    Items->Clear();
    Key = 0;
  }
  TUIStateAwareComboBox::KeyDown(Key, Shift);
}
//---------------------------------------------------------------------------
void __fastcall THistoryComboBox::SetMaxHistorySize(int AMaxHistorySize)
{
  FMaxHistorySize = AMaxHistorySize;
  while (Items->Count > FMaxHistorySize)
  {
    Items->Delete(Items->Count - 1);
  }
}
//---------------------------------------------------------------------------
void __fastcall THistoryComboBox::DoExit()
{
  TUIStateAwareComboBox::DoExit();
  if (SaveOn.Contains(soExit))
  {
    AddToHistory();
  }
}
//---------------------------------------------------------------------------
void __fastcall THistoryComboBox::DropDown()
{
  TUIStateAwareComboBox::DropDown();

  if (SaveOn.Contains(soDropDown))
  {
    AddToHistory();
  }

  int ItemWidth = GetMaxItemWidth() + ScaleByPixelsPerInch(8, this);
  int VScrollWidth = GetSystemMetricsForControl(this, SM_CXVSCROLL);
  if (Items->Count > DropDownCount)
  {
    ItemWidth += VScrollWidth;
  }

  // If dropdown fits from control's left to desktop's right edge, it's dropped to the right.
  // Otherwise it drops to the left.
  // The following limits the width do that it fits either to the right or left on the current monitor only.
  TRect MonitorRect = GetParentForm(this)->Monitor->WorkareaRect;
  TPoint ScreenPosition = ClientToScreen(TPoint());
  // VScrollWidth is used just as a convenient size of margin between monitor edge and scrollber
  int RightSpace = MonitorRect.Right - ScreenPosition.X - VScrollWidth;
  int LeftSpace = ScreenPosition.X + Width - MonitorRect.Left - VScrollWidth;
  int DropSpace = RightSpace;
  // Drop to the left only if there's considerably more space on left than right
  if (LeftSpace > RightSpace * 11 / 10)
  {
    DropSpace = LeftSpace;
  }
  ItemWidth = std::min(ItemWidth, DropSpace);

  Perform(CB_SETDROPPEDWIDTH, static_cast<unsigned int>(ItemWidth), 0);
}
//---------------------------------------------------------------------------
void __fastcall THistoryComboBox::Change()
{
  TUIStateAwareComboBox::Change();
  if (OnSetData != nullptr)
  {
    // note that ItemIndex is not reliable
    int Index = Items->IndexOf( Text );
    if (Index >= 0)
    {
      OnSetData(this, Items->Objects[Index]);
    }
  }
}
//---------------------------------------------------------------------------
void THistoryComboBox::AddToHistory()
{
  if (!Text.IsEmpty())
  {
    TObject * Data = nullptr;
    if (OnGetData != nullptr)
    {
      OnGetData(this, Data);
    }
    ::SaveToHistory(Items, Text, Data, MaxHistorySize);
    ItemIndex = 0;
  }
}
//---------------------------------------------------------------------------
void THistoryComboBox::SaveToHistory()
{
  AddToHistory();
  if (OnSaveHistory != nullptr)
  {
    OnSaveHistory(this);
  }
}
//---------------------------------------------------------------------------
int THistoryComboBox::GetMaxItemWidth()
{
  int Result = 0;
  HDC DC = GetDC(0);
  try
  {
    HGDIOBJ SaveFont = SelectObject(DC, Font->Handle);
    for (int Index = 0; Index < Items->Count; Index++)
    {
      TSize Size;
      GetTextExtentPoint32(DC, Items->Strings[Index].c_str(), Items->Strings[Index].Length(), &Size);
      if (Size.cx > Result)
      {
        Result = Size.cx;
      }
    }
    SelectObject(DC, SaveFont);
  }
  __finally
  {
    ReleaseDC(0, DC);
  }
  return Result;
}
//---------------------------------------------------------------------------
void THistoryComboBox::LoadHistory()
{
  if (!SaveOn.Empty() && !HistoryKey.IsEmpty() && HandleAllocated())
  {
    if (OnLoadHistory != nullptr)
    {
      OnLoadHistory(this);
    }
    else
    {
      Items->Clear();
    }
  }
}
//---------------------------------------------------------------------------
void __fastcall THistoryComboBox::SetHistoryKey(UnicodeString value)
{
  if (HistoryKey != value)
  {
    FHistoryKey = value;
    LoadHistory();
  }
}
//---------------------------------------------------------------------------
void __fastcall THistoryComboBox::CreateWnd()
{
  TUIStateAwareComboBox::CreateWnd();
  LoadHistory();
}
